test/e2e/e2e_suite_test.go

```diff
@@ -1,77 +1,372 @@
 
-var (
-	kubeAPIClient client.Client
-	clientset     *kubernetes.Clientset
-)
+type Framework struct {
+	Client    client.Client
+	Clientset *kubernetes.Clientset
+	GinkgoT   GinkgoTInterface
+}
+
+type Operator struct {
+	Namespeace                *corev1.Namespace
+	Deployment                *appsv1.Deployment     // meridio-operator-controller-manager
+	LeaderElectionRole        *rbacv1.Role           // meridio-operator-leader-election-role
+	ManagerRole               *rbacv1.Role           // meridio-operator-manager-role
+	LeaderElectionRoleBinding *rbacv1.RoleBinding    // meridio-operator-leader-election-rolebinding
+	ManagerRoleBinding        *rbacv1.RoleBinding    // meridio-operator-manager-rolebinding
+	ServiceAccount            *corev1.ServiceAccount // meridio-operator-controller-manager
+	ConfigMap                 *corev1.ConfigMap      // meridio-operator-manager-config
+	Service                   *corev1.Service        // meridio-operator-webhook-service
+	// certificate.cert-manager.io/meridio-operator-serving-cert
+	// issuer.cert-manager.io/meridio-operator-selfsigned-issuer
+	// mutatingwebhookconfiguration.admissionregistration.k8s.io/meridio-operator-mutating-webhook-configuration
+	// validatingwebhookconfiguration.admissionregistration.k8s.io/meridio-operator-validating-webhook-configuration
+}
+
+var fw = NewFramework()
+var trench = &meridiov1alpha1.Trench{
+	ObjectMeta: metav1.ObjectMeta{
+		Name:      trenchName,
+		Namespace: namespace,
+	},
+	Spec: meridiov1alpha1.TrenchSpec{
+		IPFamily: "DualStack",
+	},
+}
+
+var replicas = int32pointer(1)
+var attractor = &meridiov1alpha1.Attractor{
+	ObjectMeta: metav1.ObjectMeta{
+		Name:      attractorName,
+		Namespace: namespace,
+		Labels: map[string]string{
+			"trench": trenchName,
+		},
+	},
+	Spec: meridiov1alpha1.AttractorSpec{
+		VlanID:         100,
+		VlanInterface:  "eth0",
+		Replicas:       replicas, // replica of lb-fe
+		Gateways:       []string{"gateway-a", "gateway-b"},
+		Vips:           []string{"vip-a", "vip-b"},
+		VlanPrefixIPv4: "169.254.100.0/24",
+		VlanPrefixIPv6: "100:100::/64",
+	},
+}
+
+func int32pointer(i int32) *int32 {
+	var ret = new(int32)
+	*ret = i
+	return ret
+}
 
-func TestE2e(t *testing.T) {
+func TestE2E(t *testing.T) {
 	if testing.Short() {
 		t.Skip()
 	}
 	RegisterFailHandler(Fail)
-	RunSpecs(t, "E2e Suite")
+	RunSpecs(t, "E2E Suite")
 }
 
 var _ = BeforeSuite(func() {
+	deployment := fw.GetOperator(namespace)
+	require.NotNil(fw.GinkgoT, deployment)
+})
+
+var _ = AfterSuite(func() {
+	require.Equal(fw.GinkgoT, fw.OperatorPodRestarts(), int32(0))
+})
+
+func NewFramework() *Framework {
+	t := GinkgoT()
+
 	myScheme := runtime.NewScheme()
-	err := scheme.AddToScheme(myScheme)
-	Expect(err).ToNot(HaveOccurred())
-	err = apiextensions.AddToScheme(myScheme)
-	Expect(err).ToNot(HaveOccurred())
-	err = meridiov1alpha1.AddToScheme(myScheme)
-	Expect(err).ToNot(HaveOccurred())
+
+	require.NoError(t, kubescheme.AddToScheme(myScheme))
```

Comment:
Why have you used "require" (testify) and not "Expect" (gomega)

Reply:
because this function NewFramework is running outside of the RunSpecs()). Seems Expect cannot be used directly, but I found using gomega could be a workaround. Now I moved the usage of require.